sdk/common/src/main/scala/org/typelevel/otel4s/sdk/autoconfigure/Config.scala

```diff
@@ -0,0 +1,244 @@
+      string.split(",").map(_.trim).filter(_.nonEmpty).toList
+
+    private def decodeWithHint[A](
+        hint: String
+    )(decode: String => Either[ConfigurationError, A]): Reader[A] =
+      (key, properties) =>
+        Reader[String].read(key, properties).flatMap {
+          case Some(value) =>
+            Either
+              .catchNonFatal(decode(value))
+              .leftMap { cause =>
+                ConfigurationError(
+                  s"Invalid value for property $key=$value. Must be [$hint]",
+                  cause
+                )
+              }
+              .flatten
+              .map(Some(_))
+
+          case None => Right(None)
+        }
+
+  }
+
+  /** Creates a [[Config]] with the given properties. The keys of the properties
+    * will be normalized.
+    *
+    * The priorities of the values: system properties > env variables > default.
+    *
+    * @param systemProperties
+    *   the system properties (e.g. `sys.props`)
+    *
+    * @param environmentVariables
+    *   the env variables (e.g. `sys.env`)
+    *
+    * @param defaultProperties
+    *   the default properties
+    */
+  def apply(
+      systemProperties: Map[String, String],
+      environmentVariables: Map[String, String],
+      defaultProperties: Map[String, String]
+  ): Config = {
+    val default = defaultProperties.map { case (key, value) =>
+      (normalizePropertyKey(key), value)
+    }
+
+    val env = environmentVariables.map { case (key, value) =>
+      (normalizeEnvVariableKey(key), value)
+    }
+
+    val props = systemProperties.map { case (key, value) =>
+      (normalizePropertyKey(key), value)
+    }
+
+    Impl(default ++ env ++ props)
+  }
+
+  /** Creates a [[Config]] by loading properties from the env variables and
+    * system props.
+    *
+    * @param default
+    *   the default properties
+    */
+  def load[F[_]: Sync](default: Map[String, String]): F[Config] =
+    for {
+      envVars <- Env.make[F].entries
+      systemProps <- Sync[F].delay(sys.props.toMap)
+    } yield apply(systemProps, envVars.toMap, default)
+
+  private final case class Impl[F[_]](
+      properties: Map[String, String]
+  ) extends Config {
+
+    def get[A: Reader](key: String): Either[ConfigurationError, Option[A]] =
+      Reader[A].read(normalizePropertyKey(key), properties)
+
+    def getOrEmpty[A: Reader: Monoid](
+        key: String
+    ): Either[ConfigurationError, A] =
+      get[A](key).map(_.getOrElse(Monoid[A].empty))
+
+    def withOverrides(overrides: Map[String, String]): Config = {
+      val normalized = overrides.map { case (key, value) =>
+        (normalizePropertyKey(key), value)
+      }
+
+      copy(properties = this.properties ++ normalized)
+    }
+
+  }
+
+  /** Normalizes a property key by converting to lower case and replacing "-"
+    * with ".".
+    */
+  private def normalizePropertyKey(key: String): String =
+    key.toLowerCase.replace("-", ".");
```

Comment:
Using `toLowerCase` without `Locale` may lead to unexpected behavior on some systems. But Scala.js does not support `java.util.Locale`.

Since we transform keys to lowercase, we can use `CIString` instead, I guess.